Marksman/Server.fs

```diff
@@ -715,4 +721,57 @@ type MarksmanServer(client: MarksmanClient) =
 
         AsyncLspResult.success tokens
 
+    override this.TextDocumentCodeAction(opts: CodeActionParams) =
+        let state = requireState ()
+        let docPath = opts.TextDocument.Uri |> PathUri.fromString
+
+        let documentBeginning =
+            { Start = { Line = 0; Character = 0 }
+              End = { Line = 0; Character = 0 } }
+
+        let editAt rangeOpt text =
+            let textEdit =
+                { NewText = text
+                  Range = Option.defaultValue documentBeginning rangeOpt }
+
+            let mp = Map.empty.Add(opts.TextDocument.Uri, [| textEdit |])
+
+            { Changes = Some mp; DocumentChanges = None }
+
+        let toc =
+            monad' {
+                let! document = State.tryFindDocument docPath state
+                let! toc = Toc.TableOfContents.mk document.index
+
+                let rendered = TableOfContents.render toc
+                let detected = TableOfContents.detect document.text
+
+                let result = ($"{rendered}", detected)
+
+                result
+            }
+
+        let codeAction title edit =
+            { Title = title
+              Kind = Some CodeActionKind.Source
+              Diagnostics = None
+              Command = None
+              Data = None
+              IsPreferred = Some false
+              Disabled = None
+              Edit = Some edit }
+
+        let codeActions =
+            match toc with
+            | None -> Array.empty
+            | Some (render, existing) ->
+                match existing with
+                | None -> [| codeAction "Create a Table of Contents" (editAt None render) |]
+                | Some (oldTocRange) ->
+                    [| codeAction "Update Table of Contents" (editAt (Some oldTocRange) render) |]
```

Comment:
nit: *a* Table of Contents?

Reply:
I think in this case "the" will be more grammatically correct :)

Reply:
I trust you ;)